### ISSUE TITLE
Make all Class objects available to debuggers.

### ISSUE THREAD

**Issue by vvlevchenko** [oracle/graal#5251]
**make reachable classes accessible via JNI FindClass for `-g` mode**

Please include the following information:

as discussed on meeting with @fniephaus and @olpaw for expression evaluation in debugger we need get JNI reference for class 

**Describe who do you think will benefit the most.**
Debuggers might need this feature 

**Describe alternatives you've considered.**
[alternatively currently used function for prototyping](https://github.com/vvlevchenko/graal/blob/minamoto/ide-support/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/DebugHelper.java#L580)

**Comment by peter-hofer** [comment]
Note that `jclass` in JNI is merely an object handle for a `Class<?>` object, which already exist for each class.

I think adding classes to the dictionary used by `FindClass` is not the way to go because it can change the observable behavior of the application and can mask problems with the JNI configuration. Using `Heap.getLoadedClasses()` is also not great because it must do a complete heap walk to find all `Class<?>` objects in some safe context first.

I think the best solution is to enrich the type information in the debug data with the offset of the corresponding `Class<?>` object in the image heap (if it's not already there). This does not affect the rest of the application, and, when the debug information is kept in a separate file, does not affect image size either.

**Comment by adinn** [comment]
@peter-hofer @olpaw @fniephaus

> I think the best solution is to enrich the type information in the debug data with the offset of the corresponding Class<?> object in the image heap (if it's not already there)

It is relatively easy to identify the location of the class instance and pas it through to the debug info generator. However, this is not already in the Linux debug info for a good reason. There is no standard way to encode the location of a class's 'Class' instance in DWARF info because the idea of having such a metadata object modeled in the language is a Java-specific notion.

We could, of course, easily model a class pointer as a static member of the associated class e.g. make it look like as if the class included this definition:

```
class org.my.Foo : public Object {
  final public static java.lang.Class *'Foo.class' = org.my.Foo.class;
  . . .
}
```

Note that this would mean that in gdb you would reference the class as follows:

```
(gdb) p/x org.my.Foo::Foo.class
```

We might also want to expose the relevant name as a local linker symbol (indeed I think we could probably usefully make all static field names visible as local linker symbols).

Whether this is the right or best way to do this is a different question. I will consult with my colleagues to see if I can find out how this was handled with gcj.

**Comment by adinn** [comment]
@stooke Do you know whether/how this is handled in the Microsoft debugger?

**Comment by adinn** [comment]
@peter-hofer @olpaw @vvlevchenko @fniephaus

I had a better idea about this and have prototyped a solution under my own repo which I will share shortly.

The easy part of this is to expose the offset of the the class object C.class for a class C in the heap to debug info generators. That just requires adding a method to interface DebugTypeInfo which returns a positive long offset if a `C.class` exists or -1 if not. This can be done for instance classes, interfaces, array classes and primitive classes. The harder part is to model it in DWARF or whatever other debug info target is at stake.

I worked out what to do for DWARF. It is possible to advertise the Java Class instance for a class X using a DWARF constant info record. This class constant DIE is given the obvious name attribute: `Hello.class`, `java.util.List.class`, `int.class`, etc. The type attribute for these constant records would refer to the info DIE for class DynamicHub -- which prints as  'java.lang.Class` -- and the location attribute would be the address of the object in the heap. The class constans would be declared as externally visible from all compile units, not just the compile unit in which they are defined.

Note that this means the names for the constants have to be quoted when you mention them in gdb commands:

```
(gdb) ptype 'java.util.List.class'
type = class java.lang.Class : public java.lang.Object {
  private:
    java.lang.String *name;
    java.lang.Class *superHub;
    java.lang.Class *componentType;
    java.lang.Class *arrayHub;
    ...
(gdb) p *'java.util.List.class'.name->value
$16 = {
  <java.lang.Object> = {
    <_objhdr> = {
      hub = 0xaa9b98,
      idHash = 793215445
    }, <No data fields>}, 
  members of byte []:
  len = 14,
  data = 0x882a68 "java.util.List"
}
```

The class constant info record for `C.class` is embedded in the compile unit for class `C`. This automatically deals with the case where multiple occurrences of a class occur because the class exists in more than one loader. In that case name of the constant would be qualified by a namespace derived from the loader.

```
(gdb) p MyClassLoader_123456::'org.my.Foo.class'.name->value->data
$12="org.my.Foo"
(gdb) p MyClassLoader_654321::'org.my.Foo.class'.name->value->data
$12="org.my.Foo"
```

One thing I have omitted in my prototype is to add local linker symbols for the classes with corresponding linkage_name attributes in the DWARF class constant DIE produced by mangling the class name using the BFD friendly mangler. However, I think that would be a good idea as it will allow perf and valgrind to identify the constants symbolically.

**Comment by adinn** [comment]
n.b. The prototype does not seem to be correctly generating DIES for primitive class objects. I'll look into that.

**Comment by adinn** [comment]
Fixed.

**Comment by fniephaus** [comment]
Thanks, @adinn. After further discussions, we've decided not to move forward with this request. Finding all classes is just one missing piece for supporting *expression evaluation* in a debugger, finding methods for a given class the next. While that may also be relatively easy to implement, allowing the debugger to call methods is not. Instead, we're now working on JDWP-based debugging support (#5648), which could eventually provide support for *expression evaluation*.

**Comment by adinn** [comment]
@fniephaus @olpaw I am sure there are reasons for providing ways to access class and method info via JNI and JDWP but I do not see why that necessitates dropping this additional debug info? Your decision to reject this patch simply makes it harder to debug when using gdb. Committing it does not affect the viability of any JNI based solution. Why do you choose to punish gdb users?

**Comment by fniephaus** [comment]
We don't want to punish any of our users, so this is simply a misunderstanding. I believe I could've just closed this issue as fixed as your patch was already merged via #5478 (please correct me if I'm wrong). Nonetheless, the OP mentioned "expression evaluation" which needs more than this, including an IDE that can run Java code.

If your time allows, let's maybe sync in the next two weeks and discuss this and some other topics related to debug info?

**Comment by adinn** [comment]
> I believe I could've just closed this issue as fixed as your patch was already merged

@fniephaus Please accept my apologies -- I have been on PTO for the last few weeks and thought that this had been closed without merging the patch. I'm glad to hear it is in the mainline.
